src/PerpEF.sol

```diff
@@ -0,0 +1,539 @@
+    );
+
+    /**
+     * @dev event for when size is increased.
+     * @param trader: address of the trader.
+     * @param increaseInSizeInCollateralToken: amount of increased size.
+     */
+    event SizeIncreased(
+        address indexed trader,
+        uint256 increaseInSizeInCollateralToken
+    );
+
+    /// -----------------------------------------------------------------------
+    /// Constructor logic
+    /// -----------------------------------------------------------------------
+
+    /**
+     * @notice Constructor logic.
+     * @dev Also initializes the ERC20 for liquidity shares.
+     * @param collateralToken: address of the ERC20 token to be used as collateral.
+     * @param wbtc: address of the ERC20 smart contract for WBTC.
+     * @param priceOracle: address of the price oracle for WBTC.
+     * @param maxLeverage: maximum allowed leverage.
+     * @param maxUtilizationPercentage: maximum utilization percentage of liquidity in positions.
+     */
+    constructor(
+        address collateralToken,
+        address wbtc,
+        address priceOracle,
+        uint256 maxLeverage,
+        uint256 maxUtilizationPercentage
+    ) ERC20("Liquidity token PerpEF", "LTPerpEF") Ownable(msg.sender) {
+        i_collateralToken = IERC20(collateralToken);
+        i_wbtc = IERC20(wbtc);
+        i_priceOracle = AggregatorV3Interface(priceOracle);
+        s_maxLeverage = maxLeverage;
+        s_maxUtilizationPercentage = maxUtilizationPercentage;
+        TOKEN_PRECISION = 10 ** IERC20Decimals(wbtc).decimals();
+        PRICE_PRECISION = 10 ** i_priceOracle.decimals();
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Public/external state-change functions
+    /// -----------------------------------------------------------------------
+
+    /**
+     * @notice Opens a position.
+     * @param type_: position type. Either `LONG` or `SHORT`.
+     * @param collateral: amount of collateral in collateral token.
+     * @param size: size value for the position, in collateral token.
+     */
+    function newPosition(
+        PositionType type_,
+        uint256 collateral,
+        uint256 size
+    ) external {
+        // Check if the position (size / collateral) exceeds the max leverage
+        _checkIfExceedsMaxLevarage(collateral, size);
+
+        // liquidity reserves valiation - Traders cannot utilize more than a configured percentage of the deposited liquidity.
+        uint256 indexTokenPrice = getPrice();
+        uint256 sizeInIndexTokens = _convertToIndexTokens(
+            size,
+            indexTokenPrice
+        );
+        _validateLiquidityReserves(indexTokenPrice, size, sizeInIndexTokens);
+
+        // Collateral transfer
+        if (
+            i_collateralToken.allowance(msg.sender, address(this)) < collateral
+        ) {
+            revert PerpEF__NotEnoughAllowance();
+        }
+
+        s_shortOpenInterest += size;
```

Comment:
think here we need to put them into two branches based on the condition that type_=SHORT or LONG